src/algorithms/ctmrg/sparse_environments.jl

```diff
@@ -0,0 +1,265 @@
+linear map or adjoint linear map on `x` if `Val(true)` or `Val(false)` is passed, respectively.
+"""
+function (env::HalfInfiniteEnv)()  # Dense operator
+    return halfinfinite_environment(
+        env.C_1,
+        env.C_2,
+        env.E_1,
+        env.E_2,
+        env.E_3,
+        env.E_4,
+        env.ket_1,
+        env.ket_2,
+        env.bra_1,
+        env.bra_2,
+    )
+end
+function (env::HalfInfiniteEnv)(x, ::Val{false})  # Linear map: env() * x
+    return halfinfinite_environment(
+        env.C_1,
+        env.C_2,
+        env.E_1,
+        env.E_2,
+        env.E_3,
+        env.E_4,
+        x,
+        env.ket_1,
+        env.ket_2,
+        env.bra_1,
+        env.bra_2,
+    )
+end
+function (env::HalfInfiniteEnv)(x, ::Val{true})  # Adjoint linear map: env()' * x
+    return halfinfinite_environment(
+        x,
+        env.C_1,
+        env.C_2,
+        env.E_1,
+        env.E_2,
+        env.E_3,
+        env.E_4,
+        env.ket_1,
+        env.ket_2,
+        env.bra_1,
+        env.bra_2,
+    )
+end
+
+# Wrapper around halfinfinite_environment contraction using EnlargedCorners (used in ctmrg_projectors)
+function halfinfinite_environment(ec_1::EnlargedCorner, ec_2::EnlargedCorner)
+    return HalfInfiniteEnv(
+        ec_1.C,
+        ec_2.C,
+        ec_1.E_1,
+        ec_1.E_2,
+        ec_2.E_1,
+        ec_2.E_2,
+        ec_1.ket,
+        ec_2.ket,
+        ec_1.bra,
+        ec_2.bra,
+    )
+end
+
+# ------------------------------------------------------------------------
+# Methods to make environment compatible with IterSVD and its reverse-rule
+# ------------------------------------------------------------------------
+
+TensorKit.InnerProductStyle(::HalfInfiniteEnv) = EuclideanProduct()
+TensorKit.sectortype(::HalfInfiniteEnv) = Trivial
+TensorKit.storagetype(env::HalfInfiniteEnv) = storagetype(env.ket_1)
+TensorKit.spacetype(env::HalfInfiniteEnv) = spacetype(env.ket_1)
+
+function TensorKit.domain(env::HalfInfiniteEnv)
+    return domain(env.E_4) * domain(env.ket_2)[3] * domain(env.bra_2)[3]'
+end
+function TensorKit.codomain(env::HalfInfiniteEnv)
+    return codomain(env.E_1)[1] * domain(env.ket_1)[3]' * domain(env.bra_1)[3]
+end
+function TensorKit.space(env::HalfInfiniteEnv)
+    return codomain(env) ← domain(env)
+end
+function TensorKit.blocks(env::HalfInfiniteEnv)
+    return TensorKit.SingletonDict(Trivial() => env)
+end
+function TensorKit.blocksectors(::HalfInfiniteEnv)
+    return TensorKit.OneOrNoneIterator{Trivial}(true, Trivial())
+end
+function TensorKit.block(env::HalfInfiniteEnv, c::Sector)
+    return env
+end
+function TensorKit.tsvd!(f::HalfInfiniteEnv; trunc=NoTruncation(), p::Real=2, alg=IterSVD())
+    return _tsvd!(f, alg, trunc, p)
+end
+function TensorKit.MatrixAlgebra.svd!(env::HalfInfiniteEnv, args...)
+    return TensorKit.MatrixAlgebra.svd!(env(), args...)  # Construct environment densely as fallback
+end
```

Comment:
This somehow confuses me, I would not have guessed that if we dispatch to KrylovKit, we would need this TensorKit interop...